File: public/scripts/src/util/validate.ts
import {displayCalculationCard} from "./display";

const isValidBinary = (binary: string): boolean => {
    if (binary.length > 64) {
        return false;
    }

    for (let i = 0; i < binary.length; i++) {
        if (binary[i] !== "0" && binary[i] !== "1") {
            return false;
        }
    }
    return true;
};

const isValidDecimal = (decimal: number): boolean => {
    return !isNaN(decimal) && decimal < 999999999999;
};

const isValidHexadecimal = (hexadecimal: string): boolean => {
    return /^[A-F0-9]+$/.test(hexadecimal) && hexadecimal.length < 64;
};

export const validateForm = (form: Object): Array<Object> => {
    for (const key in form) {
        form[key] = form[key].replace(/\s/g, ""); // removes all whitespace from the form data
        if (form[key] === "") {
            displayCalculationCard("Please enter a valid input.");
            return [form, false];
        }

        switch (key) {
            case "binary":
                if (!isValidBinary(form[key])) {
                    displayCalculationCard("Please enter a valid binary number, and is not over 64 characters in length.");
                    return [form, false];
                }
                break;
            case "decimal":
                if (!isValidDecimal(form[key])) {
                    displayCalculationCard("Please enter a valid decimal number, and is not over 999,999,999,999.");
                    return [form, false];
                } else {
                    form[key] = parseInt(form[key]); // data must be an integer string not a float string
                }
                break;
            case "hexadecimal":
                form[key] = form[key].toUpperCase(); // hexadecimal to uppercase before it reaches the server
                if (!isValidHexadecimal(form[key])) {
                    displayCalculationCard("Please enter a valid hexadecimal number, and is not over 64 characters in length.");
                    return [form, false];
                }
                break;
            default:
                if (!isValidDecimal(form[key])) {
                    displayCalculationCard("Please enter a valid input, and is not over 999,999,999,999.");
                    return [form, false];
                }
                // percentages and total-surface-area category require a float input
                const currentPageURL = window.location.pathname;
                if (currentPageURL.indexOf("percentages") !== -1 || currentPageURL.indexOf("total-surface-area") !== -1) {
                    form[key] = parseFloat(form[key]);
                } else {
                    form[key] = parseInt(form[key]); // input is for a int route such as numbers category

                }
        }
    }
    // the function has not returned, therefore data is valid
    const spinner = document.getElementById("loading-spinner-container") as HTMLElement;
    spinner.style.opacity = "100";
    return [form, true]
};

File: app/api/mathematics/validate.go
package mathematics

import (
	"fmt"
	"strconv"
)

func validateBinary(binary string) error {
	b, err := strconv.ParseInt(binary, 2, 0)
	if err != nil || b < 1 || len(binary) > 64 {
		return fmt.Errorf("please enter a valid binary number, greater than 0 and is not over 64 "+
			"characters in length. Incorrect input: %s", binary)
	}
	return nil
}

func validatePositiveInt(numbers ...int) error {
	for _, n := range numbers {
		if n < 1 || n > 999999999999 {
			return fmt.Errorf("please enter a valid decimal number, greater than 0 and less than "+
				"1'000'000'000'000. Incorrect input: %d", n)
		}
	}
	return nil
}

func validateFloat(positiveOnly bool, floats ...float64) error {
	for _, n := range floats {
		switch positiveOnly {
		case false:
			if n > 999999999999 {
				return fmt.Errorf("please enter a valid decimal number, greater than 0 and less than "+
					"1'000'000'000'000: %f", n)
			}
		case true:
			if n < 1 || n > 999999999999 {
				return fmt.Errorf("please enter a valid decimal number, greater than 0 and less than "+
					"1'000'000'000'000: %f", n)
			}
		}
	}
	return nil
}

func validateHexadecimal(hexadecimal string) error {
	h, err := strconv.ParseInt(hexadecimal, 16, 0)
	if err != nil || h < 1 || len(hexadecimal) > 64 {
		return fmt.Errorf("please enter a valid hexadecimal number, greater than 0 and less than "+
			"1'000'000'000'000: %s", hexadecimal)
	}
	return nil
}
